drivers/mtd/rpmsgmtd.c

```diff
@@ -0,0 +1,1128 @@
+                                     uint32_t src, FAR void *priv);
+static int     rpmsgmtd_ioctl_handler(FAR struct rpmsg_endpoint *ept,
+                                      FAR void *data, size_t len,
+                                      uint32_t src, FAR void *priv);
+
+/* Functions for creating communication with remote cpu */
+
+static void    rpmsgmtd_device_created(struct rpmsg_device *rdev,
+                                       FAR void *priv_);
+static void    rpmsgmtd_device_destroy(struct rpmsg_device *rdev,
+                                       FAR void *priv_);
+static int     rpmsgmtd_ept_cb(FAR struct rpmsg_endpoint *ept,
+                               FAR void *data, size_t len, uint32_t src,
+                               FAR void *priv);
+static void    rpmsgmtd_ns_bound(struct rpmsg_endpoint *ept);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+/* Rpmsg device response handler table */
+
+static const rpmsg_ept_cb g_rpmsgmtd_handler[] =
+{
+  [RPMSGMTD_ERASE]  = rpmsgmtd_default_handler,
+  [RPMSGMTD_BREAD]  = rpmsgmtd_bread_handler,
+  [RPMSGMTD_BWRITE] = rpmsgmtd_default_handler,
+  [RPMSGMTD_READ]   = rpmsgmtd_read_handler,
+  [RPMSGMTD_WRITE]  = rpmsgmtd_default_handler,
+  [RPMSGMTD_IOCTL]  = rpmsgmtd_ioctl_handler,
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: rpmsgmtd_erase
+ *
+ * Description:
+ *   Rpmsg-mtd erase operation
+ *
+ * Parameters:
+ *   dev        - the mtd device
+ *   startblock - erase start block
+ *   nblocks    - erase block number
+ *
+ * Returned Values:
+ *   OK on success; A negated errno value is returned on any failure.
+ *
+ ****************************************************************************/
+
+static int rpmsgmtd_erase(FAR struct mtd_dev_s *dev, off_t startblock,
+                          size_t nblocks)
+{
+  FAR struct rpmsgmtd_s *priv = (FAR struct rpmsgmtd_s *)dev;
+  struct rpmsgmtd_erase_s msg;
+
+  /* Sanity checks */
+
+  DEBUGASSERT(priv != NULL);
+
+  msg.startblock = startblock;
+  msg.nblocks    = nblocks;
+
+  return rpmsgmtd_send_recv(priv, RPMSGMTD_ERASE, true, &msg.header,
+                            sizeof(msg), NULL);
+}
+
+/****************************************************************************
+ * Name: rpmsgmtd_get_geometry
+ *
+ * Description:
+ *   Rpmsg-mtd get the server mtd device geometry
+ *
+ * Parameters:
+ *   dev - the mtd device
+ *
+ * Returned Values:
+ *   OK on success; A negated errno value is returned on any failure.
+ *
+ ****************************************************************************/
+
+static int rpmsgmtd_get_geometry(FAR struct rpmsgmtd_s *dev)
```

Comment:
static int rpmsgmtd_get_geometry(FAR struct rpmsgmtd_s *dev)
{ 
  int ret;
    
  ret = nxmutex_lock(&dev->geoexcl);
  if (ret == OK && dev->geo.blocksize == 0)
    {
      /* Get the server mtd device geometry */
          
      ret = rpmsgmtd_ioctl(&dev->mtd, MTDIOC_GEOMETRY,
                           (unsigned long)&dev->geo);
    }

  nxmutex_unlock(&dev->geoexcl);
  return ret;
}